annnufan/assignment2/node.cpp

```diff
@@ -1,80 +1,91 @@
 #include <iostream>
 #include <algorithm>
 
-
 class node {
 	int key;
+	node *left_node, *right_node, *parent_node;
 public:
-	node(int x, node* l = nullptr, node* r = nullptr) : key(x), left_node(l), right_node(r){}
+	node(int x, node* l = nullptr, node* r = nullptr) : key(x), left_node(l), right_node(r), parent_node(nullptr) {
+		if (left_node != nullptr) {
+			left_node->parent_node = this;
+		}
+		if (right_node != nullptr) {
+			right_node->parent_node = this;
+		}
+	}
+
 	~node() {
 		delete left_node;
 		delete right_node;
 	}
 
-	void set_value(int x) {
-		key = x;
-	}
-
 	int get_value() {
 		return key;
 	}
 
-	node* left_node, *right_node, *parent_node;
-};
-
-bool find_key_and_print_way_to_root(int x, node* v) {
-	if (v == nullptr) {
-		return false;
-	}
-	if (x == v->get_value()) {
-		return true;
+	node* get_left() {
+		return left_node;
 	}
-	if (find_key_and_print_way_to_root(x, v->left_node) || find_key_and_print_way_to_root(x, v->right_node)) {
-		std::cout << v->get_value() << ' ';
-		return true;
+
+	node* get_right() {
+		return right_node;
 	}
-	return false;
-}
 
-void print_way_to_root_for_key(int x, node* v) {
-	if (find_key_and_print_way_to_root(x, v)) {
-		std::cout << "\n";
-	} else {
-		std::cout << "Couldn't find key\n";
+	node* get_parent() {
+		return parent_node;
 	}
-}
+};
 
-node* find_node(int x, node* v, node* p) {
+node* find_node(int key, node* v) {
 	if (v == nullptr) {
 		return nullptr;
 	}
-	v->parent_node = p;
-	if (x == v->get_value()) {
+	if (key == v->get_value()) {
 		return v;
 	}
-	node* u = find_node(x, v->left_node, v);
+	node* u = find_node(key, v->get_left());
 	if (u != nullptr) {
 		return u;
 	}
-	u = find_node(x, v->right_node, v);
+	u = find_node(key, v->get_right());
 	return u;
 }
 
+void print_path_to_root(node* v) {
+	if (v == nullptr) {
+		return;
+	}
+	while (v->get_parent() != nullptr) {
+		v = v->get_parent();
+		std::cout << v->get_value() << ' ';
+	}
+	std::cout << "\n";
+}
+
+void print_path_to_root_for_key(int key, node* root) {
+	node* v = find_node(key, root);
+	if (v == nullptr) {
+		std::cout << "Couldn't find key\n";
+		return;
+	}
+	print_path_to_root(v);
+}
+
 int count_high(node* v) {
 	if (v == nullptr) {
 		return -1;
 	}
 	int ans = 0;
-	while (v->parent_node != nullptr) {
+	while (v->get_parent() != nullptr) {
 		ans++;
-		v = v->parent_node;
+		v = v->get_parent();

```

Comment:
Tip: You can get the same result like this:
```C++
int count_high(node* v) {
    int ans = -1;
    while (v != null) {
        ans++;
        v = v->get_parent();
    }
    return ans;
}
```
Slightly shorter, but more importantly, only one call to `get_parent`.